File: LeetCode/Study Plan/Practice Problems/Medium/Longest Repeating Character Replacement/Go/longest_repeating_character_replacement.go
package main

import "fmt"

func longest_repeating_character_replacement(s string, k int) int {
	l, max_freq := 0, 0
	mp := make(map[byte]int)
	res := 0
	for r := 0; r < len(s); r++ {
		mp[s[r]]++
		max_freq = max(max_freq, mp[s[r]])
		if r-l+1-max_freq > k {
			mp[s[l]]--
			l++
		}
		res = r - l + 1
	}
	return res
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	fmt.Println(longest_repeating_character_replacement("ABAB", 2))
}

File: LeetCode/Medium/Longest Repeating Character Replacement/Java/longestRepeatingCharacterReplacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        HashMap<Character, Integer> freq = new HashMap<>();
        int maxFreq = 0;
        int res = 0;
        int j = 0;
        for (int i=0; i<s.length(); i++) {
            freq.put(s.charAt(i), freq.getOrDefault(s.charAt(i), 0)+1);
            maxFreq = Math.max(maxFreq, freq.get(s.charAt(i)));
            while(i-j-maxFreq+1 > k) {
                freq.put(s.charAt(j), freq.get(s.charAt(j))-1);
                j++;
            }
            res = Math.max(res, i-j+1);
        }
        return res;
    }
}
